Implemented numpy setxor1d

joschkabraun opened the pull request:
This is an implementation of [numpy.setxor1d](https://docs.scipy.org/doc/numpy/reference/generated/numpy.setxor1d.html) from issue #70 (and #2078). It follows the same structure as the original implementation with changes for Jax compatibility.

np.setxor1d is intrinsically hard to make jit-compatible since it depends on np.unique and its output shape depends on the values in the array. This problem will be resolved as soon as np.unique is jit-compaitible (#2760). 

The test check all one dimensional arrays and number dtypes which are not complex (as by #2760, unique does not work with complex dtypes). The test does not compile with jit since unique is not compatible with it.

joschkabraun committed:
Implemented numpy setxor1d

joschkabraun committed:
Implemented numpy setxor1d

jekbradbury commented on jax/numpy/lax_numpy.py:
```diff
@@ -3174,6 +3174,28 @@ def unique(ar, return_index=False, return_inverse=False,
+    ar1 = unique(ar1)
+    ar2 = unique(ar2)
+  else:
+    if ar1.ndim == 0:
```
This will fail if an input is a Python scalar rather than a NumPy/JAX scalar.

jekbradbury commented on jax/numpy/lax_numpy.py:
```diff
@@ -3174,6 +3174,28 @@ def unique(ar, return_index=False, return_inverse=False,
+  if aux.size == 0:
+      return aux
+
+  aux.sort()
```
This relies on the in-place behavior of NumPy `sort`; you want to use something like `aux = sort(aux)` for JAX compatibility.

jacobjinkelly committed:
add erf inv rule

erf_inv rule not working

works up to order 2

erf inv rule

use np for now

actually use np for now

joschkabraun commented:
Okay, thank you for your feedback! I’m currently working on it and will finish it until the weekend!

joschkabraun committed:
Incorporated changes

joschkabraun committed:
Incorporated changes

joschkabraun commented:
The tests are running smooth on my local machine but I am getting errors here which seem to be outside of my code. Can you help me with them?

jakevdp commented:
As with #2941, you need to rebase on master, fix tests locally, and then force push to update this PR. Please let me know if you need help with that process.

joschkabraun committed:
Merge pull request #1 from google/master

update own master

joschkabraun committed:
rebased branch

joschkabraun committed:
rebased branch

joschkabraun committed:
changed onp to np

joschkabraun committed:
changed onp to np

joschkabraun committed:
corrected style

joschkabraun committed:
corrected style

joschkabraun committed:
corrected indent

joschkabraun committed:
corrected indent

jakevdp commented:
``rng_factory`` now requires ``self.rng()`` as the first argument - take a look at how other test routines use it.

joschkabraun committed:
adapted tests for change in creation of rng

joschkabraun committed:
adapted tests for change in creation of rng

joschkabraun commented:
Okay, thank you!

joschkabraun commented:
Running the tests here gives me the error "NotImplementedError: Numpy function <function setxor1d at 0x7fae7fd63ea0> not yet implemented". Do you know how that can happen?

jakevdp commented:
> Running the tests here gives me the error "NotImplementedError: Numpy function <function setxor1d at 0x7fae7fd63ea0> not yet implemented"

You'll need to add the function to the top-level imports here: https://github.com/google/jax/blob/e48a4e012bd253428e3bb0dd4ac5b163439a5270/jax/numpy/__init__.py#L20-L59

joschkabraun committed:
changed import for numpy

joschkabraun committed:
changed import for numpy

joschkabraun commented:
Okay, thank you!

joschkabraun committed:
updated tests to not use bfloat16

joschkabraun committed:
updated tests to not use bfloat16

jakevdp commented:
Looks like the remaining error is a dtype mismatch in x64 mode. I'd suggest passing ``check_dtypes=False`` in the test function.

joschkabraun committed:
adapted test

joschkabraun committed:
adapted test

joschkabraun commented:
Okay, thank you!

jakevdp commented:
Looks good - please also remove the two ``.orig`` files that were added along the way.

joschkabraun committed:
remove lax_numpy.py.orig

joschkabraun committed:
remove lax_numpy.py.orig

joschkabraun committed:
remove lax_numpy_test.py.orig

joschkabraun committed:
remove lax_numpy_test.py.orig

joschkabraun commented:
Okay, this has happened now!

jakevdp commented on jax/numpy/lax_numpy.py:
```diff
@@ -3320,6 +3319,24 @@ def unique(ar, return_index=False, return_inverse=False,
+    ar2 = unique(ar2)
+  else:
+    ar1 = asarray(ar1)
+    ar2 = asarray(ar2)
```
We should not call `asarray`; valid inputs will be properly handled by `concatenate`.

jakevdp commented on jax/numpy/lax_numpy.py:
```diff
@@ -3320,6 +3319,24 @@ def unique(ar, return_index=False, return_inverse=False,
+  
+  aux = concatenate((ar1, ar2))
+  if aux.size == 0:
+      return aux
```
Use two-space indent to match the rest of the file.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -1262,6 +1262,23 @@ def testUnique(self, shape, dtype, return_index, return_inverse, return_counts):
+    rng = jtu.rand_default(self.rng())
+    args_maker = lambda: [rng(shape1, dtype1), rng(shape2, dtype2)]
+    np_fun = lambda ar1, ar2: np.setxor1d(ar1, ar2)
+    jnp_fun = lambda ar1, ar2: jnp.setxor1d(ar1, ar2)
```
Since there's no complilation test, we can simplify to
```
np_fun = np.setxor1d
jnp_fun = jnp.setxor1d
```

joschkabraun committed:
removed asarray and corrected style

joschkabraun committed:
removed asarray and corrected style

joschkabraun committed:
simplified code in test

joschkabraun committed:
simplified code in test

joschkabraun commented:
Okay, incorporated changes!

jakevdp commented:
I just looked a bit closer at the numpy implementation: this seems to be essentially a line-for-line copy of the `setxor1d` code in numpy. That's fine, but for licensing purposes we'll need to move the implementation into ``jax/third_party/numpy`` before we can merge it.

joschkabraun commented:
Okay, that makes sense! Should I create a file `lax_numpy.py` similar to the structure in `jax/numpy` as there is only `linalg.py` in `jax/third_party/numpy`?

jakevdp commented:
Sure, you can create `jax/third_party/numpy/lax_numpy.py`, reflecting the structure of the main package.

joschkabraun committed:
Moved code to third party

joschkabraun committed:
Moved code to third party

jakevdp commented:
Looks great! The last thing we need is a rebase against master to fix the conflicts that have cropped up over the last few days. Please let me know if you need tips on how to do that.

juliuskunze committed:
Remove special case for scan masking

The scan masking rule is now simplified to a standard masking rule, and the special case handling in masking.py has been removed.

juliuskunze committed:
Minor

juliuskunze committed:
Remove unused scan shape rule

juliuskunze committed:
Fix dtype, minor

froystig committed:
de-duplicate constants when lowering to XLA

Co-authored-by: Matthew Johnson <mattjj@google.com>

froystig committed:
reuse commonly-typed input/output values when joining partially evaluated conditional branches

froystig committed:
factor autodiff and vmap tests out from lax_test

fehiepsi committed:
allow scalar input in poisson sampler

apaszke committed:
Skip known invars and outvars in JaxprTrace.process_call (#3242)

mattjj committed:
Merge pull request #3418 from fehiepsi/poissontype

allow scalar input in poisson sampler

hawkinsp committed:
Bump XLA version. (#3424)

Update jaxlib release notes.

mattjj committed:
add docstring to ravel_pytree

mattjj committed:
deflake

mattjj committed:
print warning when doing jit-of-pmap

mattjj committed:
Merge pull request #3425 from google/document-ravel-pytree

add docstring to ravel_pytree

dbudden committed:
Propagate raw __name__ and __doc__ of functions wrapped by jit and sharded_jit.

mattjj committed:
fix and better explain complex JVPs / VJPs

fixes #3433

mattjj committed:
fix typos

mattjj committed:
Merge pull request #3434 from google/autodiff-cookbook-complex-bug-fixes

fix and better explain complex JVPs / VJPs

mattjj committed:
add jax.numpy.concatenate(..., axis=None) support

fixes #3419

mattjj committed:
fix vmap-of-pmap mapped_invars logic bug

fixes #3399

This crept in via #1959, but more importantly it shows we don't have
good test coverage here!

mattjj committed:
Merge pull request #3436 from google/issue3419

add jax.numpy.concatenate(..., axis=None) support

mattjj committed:
Merge pull request #3439 from google/issue3399

fix vmap-of-pmap mapped_invars logic bug

shoyer committed:
Document valid enum values for precision. (#3441)

This is a little tricky to figure out otherwise.

gnecula committed:
Added jax2tf test about primitive coverage (#3420)

gnecula committed:
[jax2tf] Fixed the handling of `core.unit` in control-flow primitives. (#3432)

* Fixed the handling of `core.unit` in control-flow primitives.

* Remove add_any from the list of unimplemented tf ops

gnecula committed:
Fixed bug in argument type promotion for dynamic_update_slice (#3427)

Also added tests for lax.slice, lax.dynamic_slice and
lax.dynamic_update_slice.

gnecula committed:
[jax2tf] separated primitives_test.py (#3445)

* Started reorg

* [jax2tf] separated the primitives_test.py

gnecula committed:
[jax2tf] fix the too-early use of tf.constant (#3446)

If I leave it at top-level, I get test failures about missing platform Host

apaszke committed:
Initial version of invertible AD implementation (#3232)

This is a prototype implementation of the memory-efficient VJP method
for invertible function. The general idea is that thanks to
invertibility, we don't have to memoize any intermediate primal values,
but can simply reconstruct them in lock-step with gradient computation.
The API is such that the only thing a user has to do, is decorate a
function with `@invertible`, which will make AD apply the more efficient
transpose than usual.

The current version is expressive enough to support e.g. the Reversible
ResNet, but there are still some caveats:
- The definition of "invertible" function is a one that produces a jaxpr
  that can be inverted correctly if only we iterate over its equations
  in reverse. This is a bit strict, because users generally don't have
  too much control over that, and there are functions that produce
  jaxprs which will be treated as invertible when one topological
  ordering of equations is used, while they will be considered
  non-invertible for other valid orderings.
- It doesn't follow the usual jvp + transpose path, and it turns out
  that zero argument pruning in JVPTrace makes it pretty much impossible
  to implement correctly.
- `custom_ivjp` is an initial-style primitive.
- Invertible reverse-mode implementation (`rev_backward_pass`) assumes
  that all the VJPs of primal primitives are jittable (not sure if
  that's a problem, but worth pointing out).
- Not having a dedicated linearization pass makes the JVP of
  `custom_ivjp` inefficient if it is being staged out.

mattjj committed:
roll back of #3439 while we debug internal fails

mattjj committed:
Merge pull request #3448 from google/roll-back-3439

roll back of #3439 while we debug internal fails

mattjj committed:
add systematic tests for vmap-of-pmap

fixes #3440

Also re-applies the fix in #3439 (i.e it rolls-back the rollback PR #3448) because we're now confident it's correct (and some internal tests are buggy).

mattjj committed:
deflake

mattjj committed:
Merge pull request #3449 from google/issue3440

add systematic tests for vmap-of-pmap

jakevdp committed:
Fix jnp.arange(x) for fractional x (#3453)

jacobjinkelly committed:
add int pow rule

mattjj committed:
simplify remat partial eval parameterization

The main win here is reducing the number of arguments for the function
that parameterizes _remat_partial_eval (so it can be used both with
remat and invertible ad features).

I also included a fix to _remat_partial_eval that is needed in #3370,
though I don't think it's needed on master. It was easier to include the
fix now.

Both these changes made rebasing #3370 easier!

fehiepsi committed:
add scipy.special.zeta, scipy.special.polygamma (#3385)

mattjj committed:
deflake

froystig committed:
jaxpr pretty-print: wrap equation RHS when the LHS is long

mattjj committed:
Merge pull request #3426 from google/warn-jit-of-pmap

print warning when doing jit-of-pmap

mattjj committed:
Merge branch 'master' into changelist/316251368

mattjj committed:
Merge pull request #3429 from BuddenD/changelist/316251368

Propagate raw __name__ of functions wrapped by jit and sharded_jit.

mattjj committed:
Merge pull request #3455 from jacobjinkelly/erf_inv

add erf_inv rule

mattjj committed:
Merge pull request #3456 from jacobjinkelly/int_pow

Jet rule for `integer_pow`

mattjj committed:
Merge pull request #3396 from JuliusKunze/simplify-scan-mask

Remove special case for scan masking

mattjj committed:
added reviewer suggestion

mattjj committed:
deflake

mattjj committed:
Merge pull request #3459 from google/simplify-remat-partial-eval

simplify remat partial eval parameterization

froystig committed:
update jaxlib version in readme

mattjj committed:
fix a bug from #3459 (#3466)

mattjj committed:
update tpu readme not to pin jax / jaxlib versions (#3467)

jacobjinkelly committed:
add jet primitives, refactor tests (#3468)

Co-authored-by: Jesse Bettencourt <jessebett@cs.toronto.edu>

gnecula committed:
Fixed bug with silent TF using 64-bit operations in 32-bit mode. (#3471)

We already cast the NumPy arrays to 32-bit, but this was not happening
if the input was tf.Variable or tf.constant.

In the process I have added some invariant checks, if core.skip_checks is
False, which it always is in testing.

Added also some sanity checking or arguments.

gnecula committed:
[jax2tf] Added special case for tf.pad. (#3462)

Fixed lax_reference.pad to handle lax.pad with negative edge padding.

ashridh committed:
Add polyder numpy function (#3403)

mattjj committed:
use original numpy for shape calculations (#3474)

cf. #3453

jakevdp committed:
Add experimental precision doubling transform (#3465)

This PR adds an experimental precision doubling transform, following the basic approach outlined in Dekker 1971 ([pdf](http://csclub.uwaterloo.ca/~pbarfuss/dekker1971.pdf)). When this transform is applied, the number of significant bits is approximately doubled compared to the base operation.

 Simple demo:
```python
In [1]: import jax.numpy as jnp                                                                                        

In [2]: from jax.experimental.doubledouble import doubledouble                                                         

In [3]: def f(a, b): 
   ...:     return a + b - a 
   ...:                                                                                                                

In [4]: f(1E20, 1.0)  # float64 loses precision
Out[4]: 0.0

In [5]: g = doubledouble(f)(1E20, 1.0)
Out[5]: DeviceArray(1., dtype=float64)
```

This initial experiment supports basic arithmetic operators and inequalities.

jakevdp committed:
disable new jax2tf TPU test (#3476)

hawkinsp committed:
Attach source info to Jaxpr equations. (#3421)

* Attach source info to Jaxpr equations.

Example:
```
In [1]: import jax, jax.numpy as jnp
In [2]: def f(x, y):
   ...:    z = jax.numpy.cos(x)
   ...:    z = z * jax.numpy.tanh(y)
   ...:    return z + 2
   ...:

In [3]: jax.make_jaxpr(jax.value_and_grad(f))(7., 9.)
Out[3]:
{ lambda  ; a b.
  let c = cos a  [<ipython-input-2-5d59f71cb65d>:2 (f)]
      d = tanh b  [<ipython-input-2-5d59f71cb65d>:3 (f)]
      e = mul c d  [<ipython-input-2-5d59f71cb65d>:3 (f)]
      f = add e 2.0  [<ipython-input-2-5d59f71cb65d>:4 (f)]
      g = mul 1.0 d  [<ipython-input-2-5d59f71cb65d>:3 (f)]
      h = neg g  [<ipython-input-2-5d59f71cb65d>:2 (f)]
      i = sin a  [<ipython-input-2-5d59f71cb65d>:2 (f)]
      j = mul h i  [<ipython-input-2-5d59f71cb65d>:2 (f)]
  in (f, j) }

In [7]: print(jax.xla_computation(jax.value_and_grad(f))(7., 9.).as_hlo_module().to_string())
HloModule xla_computation_f__4.15

ENTRY %xla_computation_f__4.15 (parameter.1: f32[], parameter.2: f32[]) -> (f32[], f32[]) {
  %constant.3 = pred[] constant(false)
  %parameter.1 = f32[] parameter(0)
  %cosine.4 = f32[] cosine(f32[] %parameter.1), metadata={op_type="cos" op_name="xla_computation(f)/cos" source_file="<ipython-input-2-5d59f71cb65d>" source_line=2}
  %parameter.2 = f32[] parameter(1)
  %tanh.5 = f32[] tanh(f32[] %parameter.2), metadata={op_type="tanh" op_name="xla_computation(f)/tanh" source_file="<ipython-input-2-5d59f71cb65d>" source_line=3}
  %multiply.6 = f32[] multiply(f32[] %cosine.4, f32[] %tanh.5), metadata={op_type="mul" op_name="xla_computation(f)/mul" source_file="<ipython-input-2-5d59f71cb65d>" source_line=3}
  %constant.7 = f32[] constant(2), metadata={op_type="add" op_name="xla_computation(f)/add" source_file="<ipython-input-2-5d59f71cb65d>" source_line=4}
  %add.8 = f32[] add(f32[] %multiply.6, f32[] %constant.7), metadata={op_type="add" op_name="xla_computation(f)/add" source_file="<ipython-input-2-5d59f71cb65d>" source_line=4}
  %constant.9 = f32[] constant(1), metadata={op_type="mul" op_name="xla_computation(f)/mul" source_file="<ipython-input-2-5d59f71cb65d>" source_line=3}
  %multiply.10 = f32[] multiply(f32[] %constant.9, f32[] %tanh.5), metadata={op_type="mul" op_name="xla_computation(f)/mul" source_file="<ipython-input-2-5d59f71cb65d>" source_line=3}
  %negate.11 = f32[] negate(f32[] %multiply.10), metadata={op_type="neg" op_name="xla_computation(f)/neg" source_file="<ipython-input-2-5d59f71cb65d>" source_line=2}
  %sine.12 = f32[] sine(f32[] %parameter.1), metadata={op_type="sin" op_name="xla_computation(f)/sin" source_file="<ipython-input-2-5d59f71cb65d>" source_line=2}
  %multiply.13 = f32[] multiply(f32[] %negate.11, f32[] %sine.12), metadata={op_type="mul" op_name="xla_computation(f)/mul" source_file="<ipython-input-2-5d59f71cb65d>" source_line=2}
  ROOT %tuple.14 = (f32[], f32[]) tuple(f32[] %add.8, f32[] %multiply.13)
}
```

Co-authored-by: Matthew Johnson <mattjj@google.com>

skye committed:
Use myst to parse markdown docs, and convert a page from rst to markdown. (#3477)

Also sets the minimum sphinx version to 2.1.

skye committed:
Remove documentation presubmit action now that the RTD presubmit is enabled. (#3480)

jekbradbury committed:
Avoid tuple psum in pmean (#3479)

We have an optimization to avoid doing flops (or rather, communication) for `psum(1)` (instead we look up the axis size at trace time). But although `pmean(x)`, meaning `psum(x) / psum(1)`, is likely the most common user of `psum(1)`, it doesn't actually trigger this optimization right now because it's implemented as `psum((x, 1))` and `bind` lifts the 1 into the same `JaxprTrace` as `x` rather than letting the psum impl rule see it. The major reason for using tuple psum—providing a fixed order to avoid multihost GPU deadlocks—doesn't apply here because we don't expect the `psum(1)` to lower to an actual XLA AllReduce.

gnecula committed:
[jax2tf] Control on which TF device the TF converted code runs. (#3482)

* [jax2tf] Control on which TF device the TF converted code runs.

The device should match the device on which the JAX code runs,
otherwise the numerical comparisons don't make sense.

* This code requires some build-file changes in Google3 to properly
  run on GPU and TPU.

* Fix warnings

gnecula committed:
Disable the workaround to prevent expansion of type aliases (#3485)

* Disable the workaround to prevent expansion of type aliases

* Fix flake

jakevdp committed:
Improve dtype test coverage for random_test (#3254)

srvasude committed:
Add random_gamma_grad and use in jax.random.gamma (#3281)

hawkinsp committed:
Fix test failure in jax2tf due to conflicting merges. (#3492)

skye committed:
Bump jaxlib version to 0.1.49 and update WORKSPACE (#3495)

skye committed:
Allow specifying both `devices` and `axis_size` to pmap. (#3475)

This allows providing custom device assignments to nested pmaps or pmap-of-sharded_jit when running on a multi-host platform.

jakevdp committed:
implement jax.random.choice (#3463)

jakevdp committed:
support stacked doubling (#3490)

8bitmp3 committed:
Fix a link rendering to Autograd's reverse-mode Jacobian method (#3501)

8bitmp3 committed:
Another small fix of the link rendering in the Autodiff Cookbook - vmap transformation (#3502)

mattjj committed:
pin numpy version in setup.py to avoid warnings (#3509)

8bitmp3 committed:
Update JAX FAQ.rst, jax.device_put (#3496)

mattjj committed:
fix jet typo: jnp not np (#3507)

skye committed:
Revert "Remove documentation presubmit action now that the RTD presubmit is enabled. (#3480)" (#3497)

This (partially) reverts commit 68dcbdd1189cd938bf6023e4e1efaf64c71629aa.

@hawkinsp points out this was running doctest, which the RTD presubmit doesn't do AFAIK. We still don't build the docs here, as the RTD presubmit takes care of that.

Co-authored-by: Stephan Hoyer <shoyer@google.com>

IgorWilbert committed:
Fix typo understanding jaxprs page on readthedocs (#3513)

jakevdp committed:
Add private class wrapper for double-double arithmetic (#3521)

skye committed:
Add instructions for how to use the TensorBoard profiler to the profiling docs. (#3481)

mattjj committed:
remove the lower_fun default multiple_results=True (#3524)

NeilGirdhar committed:
Fix typo: np.bool -> np.bool_ (#3525)

Replaced np.bool (which is just bool) with np.bool_, which is numpy's
Boolean type.

tomhennigan committed:
Cast int8 to bool for lax.not in jax2tf. (#3519)

jakevdp committed:
Add jax.scipy.signal.detrend (#3516)

mattjj committed:
fix an issue with newer versions of pytype (#3526)

clemisch committed:
Remove unnecessary static_argnum in np.gradient (#3512)

thomaskeck committed:
Adds boolean support for bitwise not and unittests for boolean support on logical operations. (#3483)

Co-authored-by: Thomas Keck <thomaskeck@google.com>

hawkinsp committed:
Fix memory leak when no axis is provided to pmap. (#3394)

* Fix memory leak when no axis is provided to pmap.

* Work around flake8 false positive.

Co-authored-by: Matthew Johnson <mattjj@google.com>

majnemer committed:
Fix lax_reference's round for edge case inputs (#3309)

* Fix lax_reference's round for edge case inputs

- round(8388609) would compute trunc(8388609 + 0.5) == 8388610.
  Fix this by not modifying sufficiently inputs.
- round(0.499999970198) would compute trunc(0.499999970198 + 0.5) == 1.0
  Fix this by explicitly special casing the first float before 0.5.

* Add explicit casts to lax_reference implementation of `round`.

Co-authored-by: Peter Hawkins <phawkins@google.com>

jakevdp committed:
lax.sort: allow any sequence of Arrays, not just tuples (#3367)

mattjj committed:
refactor call primitives, simpler param processing (#3491)

mattjj committed:
add back a full_lower, dropped in #3491 (#3530)

mattjj committed:
avoid value-based error check in random.choice (#3531)

dpfau committed:
Support `b` and `return_sign` in scipy.special.logsumexp (#3488)

clemisch committed:
Add np.unwrap (#3527)

joaogui1 committed:
Future warning about lists and tuples (#3369)

nova77 committed:
Fixed issue with the 'compress' method of 'jnp.ndarray' (#3537)

* Fixed issue with the 'compress' method of 'jnp.ndarray' behaving differently than the free function

* Added test

yurodiviy committed:
Implement nan-quantile/median/percentile (#2397)

* Implement nan-quantile/median/percentile

* Fix flake8 error.

Co-authored-by: Peter Hawkins <phawkins@google.com>

hawkinsp committed:
Add missing functions to documentation. (#3538)

ashridh committed:
Add np.diag_indices_from (#3500)

hawkinsp committed:
Merge quantile and nanquantile implementations. (#3539)

Allows more code sharing, and adds support for other interpolation modes to nanquantile().

hawkinsp committed:
Restrict .compress tests to arrays only. (#3540)

Fixes test failures with 100 generated cases.

hawkinsp committed:
Fix test failures due to RuntimeWarning in variance tests. (#3541)

hawkinsp committed:
Avoid direct type/dtype comparisons to fix NumPy 1.19 deprecation war… (#3543)

* Avoid direct type/dtype comparisons to fix NumPy 1.19 deprecation warnings.

* Pin a newer tf-nightly to fix jax2tf tests for NumPy 1.19.0

JonyEpsilon committed:
Clarify docstrings regarding usage of static arguments in jit and vmap. (#3484)

The docstring for pmap does not currently mention that any "non-data"
arguments need to be indicated in `static_broadcasted_argnums`. This
commit updates the docs to parallel those for `jax.jit` which does
explain this. Additionally, a remark is added to the `static_*` argument
descriptions on both jit and pmap, so that this point can be understood
without reading the whole docstring.

hawkinsp committed:
Remove warning suppression for tuple and list arguments to reductions. (#3545)

Fix callers.

hawkinsp committed:
Fix failure in PyTorch array interoperability test in non-x64 mode. (#3546)

IgorWilbert committed:
Fix typos pytrees page on readthedocs (#3548)

froystig committed:
raise a custom error in jaxpr checker

mattjj committed:
add remat docstring (#3542)

* add remat docstring

first part of addressing #3314

jakevdp committed:
Add ability to specify individual test targets

jakevdp committed:
fix missing imports

hawkinsp committed:
Fix eigh JVP to ensure that both the primal and tangents of the eigen… (#3550)

* Fix eigh JVP to ensure that both the primal and tangents of the eigenvalues are real.

Add test to jax.test_util.check_jvp that ensure the primals and both the primals and tangents produced by a JVP rule have identical types.

* Cast input to static indexing grad tests to a JAX array so new type check passes.

mattjj committed:
add remat to top-level docs (#3554)

j-towns committed:
Fix lazy broadcast issue (#3536)

froystig committed:
introduce custom typecheck rules, implement them for cond and scan

hawkinsp committed:
Make CUDA wheels manylinux2010 compliant, add CUDA 11, drop CUDA 9.2 (#3555)

* Use dynamic loading to locate CUDA libraries in jaxlib.

This should allow jaxlib CUDA wheels to be manylinux2010 compliant.

* Tag CUDA jaxlib wheels as manylinux2010.

Drop support for CUDA 9.2, add support for CUDA 11.0.

* Reorder CUDA imports.

hawkinsp committed:
Update XLA in preparation for a new jaxlib release (0.1.50). (#3560)

froystig committed:
limit jaxpr context in typechecker error messages

mattjj committed:
allow closures for odeint dynamics functions (#3562)

* allow closures for odeint dynamics functions

fixes #2718, #3557

* add tests for odeint dynamics closing over tracers

mattjj committed:
update version and changelog for pypi (#3564)

mattjj committed:
removed stale faq entries (#3565)

jakevdp committed:
Fix logsumexp test (#3563)

Previously, this test could generate an axis out of bounds error with large num_generated_cases. (discovered in the process of testing #3561)

mattjj committed:
fix error when doing forward-mode of odeint (#3566)

fixes #3558

hawkinsp committed:
Update docker script for CUDA 11. (#3571)

hawkinsp committed:
Fix test failures on GPU. (#3572)

hawkinsp committed:
Update README for jaxlib 0.1.50 release. (#3574)

nova77 committed:
Added missing is_stable argument to lax.sort (#3553)

mattjj committed:
ensure lax.reduce monoid test uses original numpy (#3573)

malcolmreynolds committed:
Clarify docs on jax.lax.cond. (#3569)

hawkinsp committed:
Add a heap profiler API and document it. (#3576)

frederikwilde committed:
Correction a typo of the period of the PRNG. (#3578)

froystig committed:
detailed assertion in profiler test

froystig committed:
lax.sort: stable by default

gnecula committed:
[jax2tf] Add special case for translation of lax.gather to tf.gather. (#3486)

Also adds more tests for conversion for gather.

hawkinsp committed:
Implement DeviceArray.tobytes(). (#3585)

Move tolist() implementation onto DeviceArray; there's no reason to populate it dynamically.

hawkinsp committed:
Incorporate a few recent NumPy API extensions. (#3586)

sriharikrishna committed:
Issue1635 expm frechet (#2062)

* Implement Frechet derivatives for expm.

* Update expm to use the current custom gradients API.

Make some stylistic fixes.

Co-authored-by: Peter Hawkins <phawkins@google.com>

Commit:
add tests

Commit:
add tests

Commit:
numpy doesn't have a cummax, but it does have maximum.accumulate - curious api choice

Commit:
fix typo

Commit:
ignore cummax/cummin in jax2tf

Commit:
add cummax/min to __init__.py

Commit:
add trailing p to excluded names

Commit:
don't require passing identity value.  It isn't the initial value - identity is required for implementation correctness

Commit:
use correct iinfo finfo names

Commit:
onp.finfo -> jnp.finfo for bfloat16

Commit:
Add default value of None for unit in TPU impl of cummax/cummin

Commit:
Also update custom tpu rule to set unit correctly based on dtype

Commit:
account for different names of reducer in tpu function

Commit:
fix other branch

Commit:
remove unit and determine automatically for all ops

Commit:
add generic reducer primitive generator and replace prod/max/min with it.

Commit:
actually return the primitive

Commit:
consolidate jvp rule definitions

Commit:
remove now unneeded type def

Commit:
consolidate tests

Commit:
remove trailing whitespace

mattjj committed:
odeint: don't hoist non-differentiable consts (#3587)

fixes #3584

This could use further revision! Left a todo.

The issue is that in #3562 we started closure-converting the dynamics
function (by tracing it to a jaxpr up-front) so as to handle closed-over
constants with respect to which we want to differentiate the odeint
call. But if the dynamics function closes over integer-valued constants,
then we can no longer call `vjp` on the closure-converted function
without getting an error.

One fix would be to support (trivial) differentiation with respect to
integer-valued inputs. That would work if we supperss the error message
for integer-valued inputs in `vjp` and add a trivial tangent space
for integer-valued arrays. Since that's potentially a further-reaching
change, this commit instead just applies a local fix to avoid adding
integer-valued inputs to the dynamics function by adapting the
closure-conversion code.

mattjj committed:
bump version for pypi (#3590)

Commit:
floating point identity to inf

gnecula committed:
[jax2tf] Add support for custom JVP/VJP (#3581)

* [jax2tf] Add support for custom JVP/VJP

The custom VJP/JVP in initial style control-flow primitives make
use of special custom_jvp_call_jaxpr primitives

* Fix flake

ekelsen committed:
Merge branch 'master' into master

Commit:
change ending tick mark style

Commit:
Merge branch 'master' of github.com:ekelsen/jax

jakevdp committed:
Use re.search and include test class name

Commit:
add support bool identity values

Commit:
bool -> bool_ for reasons that make no sense, (bool used to be any?!)

froystig committed:
add source info to jaxpr typechecking messages

jakevdp committed:
Add ability to specify individual test targets via a regex (#3549)

* Add ability to specify individual test targets

* fix missing imports

* Use re.search and include test class name

jakevdp committed:
Merge pull request #3588 from ekelsen/master

Add cummax and cummin

jekbradbury committed:
Use precision=HIGHEST in expm repeated squaring (#3601)

* Use precision=HIGHEST in expm repeated squaring

This will improve the output accuracy on TPUs, and doesn't currently affect other platforms.

Also remove a spurious duplicate line.

* Also use HIGHEST precision in Pade approximants

* Update linalg.py

* Update linalg.py

jakevdp committed:
Cleanup: remove some test interdependence (#3600)

tomhennigan committed:
Retain original docstring when vmap'ing functions. (#3592)

gnecula committed:
[jax2tf] Fix tf.gather handling of out-of-bounds indices (#3604)

JAX and XLA clamp out-of-bounds indices, while tf.gather aborts. We
ensure that when we rewrite lax.gather to tf.gather, we use XLA.

gnecula committed:
[jax2tf] Re-enable JIT (#3605)

JIT was disabled because initially we did not have support for
control-flow primitives. Then we found out a few other unimplemented
features that prevented control-flow primitives for working properly:
handling of units (#3432), casts (#3471, #3427), custom VJP/JVP (#3581).
All these are fixed now, and it seems to be time to re-enable JIT.

hawkinsp committed:
Fix shape rule for lax.pad for input dimensions of size 0. (#3608)

jakevdp committed:
Rework type support for lax cumulative reductions (#3609)

froystig committed:
add jit-consistency checks to a few lax_numpy tests

mattjj committed:
fix reduction repeated axis error (#3618)

* fix reduction repeated axis error

* deflake

mattjj committed:
improve vmap axis spec structure mismatch errors (#3619)

* improve vmap axis spec structure mismatch errors

fixes #3613

* deflake

hawkinsp committed:
Reimplement argmin/argmax using a single pass variadic reduction. (#3611)

hawkinsp committed:
Update XLA. (#3623)

yashk2810 committed:
Add try; except to cells that fail and print out the error message so that the notebook can execute end-to-end (#3624)

mattjj committed:
skip some ode tests on gpu for speed (#3629)

yashk2810 committed:
Install scikit-image and add try;except around from google.colab import files import statement as google.colab is not available locally and in kokoro (#3626)

mattjj committed:
skip gamma tests on tpu for compilation speed (#3631)

skye committed:
Enable XLA SPMD partitioning by default. (#3151)

This option is needed for sharded_jit. Future APIs may use MPMD partitioning instead.

mattjj committed:
fix bug in categorical test, disable #3611 on tpu (#3633)

* fix bug in categorical test, disable #3611 on tpu

Disabling #3611 on TPU pending a TPU compilation bug.

* unskip a test

jakevdp committed:
Add CI tests for numpy 1.16.4 (#3634)

jakevdp committed:
Add test of jax.numpy vs numpy call signatures (#3628)

gnecula committed:
[jax2tf] Update the tf-nightly version to 2020-07-01 (#3635)

gnecula committed:
Release jaxlib 0.1.51 (#3636)

gnecula committed:
Change the version of tensorflow, the one in jaxlib 0.1.51 does not build (#3637)

gnecula committed:
Updated minimum jaxlib to 0.1.51 (#3642)

mattjj committed:
fix prng key reuse in differential privacy example (#3646)

fix prng key reuse in differential privacy example

romanngg committed:
Add support for 0d transpose convolution (#3643)

* Allow 0d transpose convolution

* Add test for 0d conv transpose

* remove whitespace

gnecula committed:
[jax2tf] Add support for first-order AD to converted functions (#3593)

* [jax2tf] Add support for first-order AD to converted functions

On conversion, optionally, we convert the jax.vjp of the converted
function and set the result as tf.custom_gradient.

* Only first-order diff is supported for now.
* Add more tests, including for round-tripping through SavedModel.

* Minor whitespace fix

mattjj committed:
make psum transpose handle zero cotangents (#3653)

make psum transpose handle zero cotangents

fixes #3651

mattjj committed:
improve concreteness error message for nn.one_hot (#3656)

* improve nn.one_hot and jax.numpy.arange errors

fixes #3654

* deflake

* debug

gnecula committed:
[jax2tf] Fix out-of-bounds errors for dynamic slices (#3649)

* [jax2tf] Fix out-of-bounds errors for dynamic slices

JAX and XLA clamp dynamic out-of-bounds indices. Reproduce that
behavior by forcing the use of XLA for tf.dynamic_slice.

We choose this solution for now for consistency with how we handle
gather. We may decide to switch to explicitly clamping indices.

Fixes: #3621

IgorWilbert committed:
Adds information about iterator in fori_loop in The Sharp Bits documentation (#3632)

tomhennigan committed:
Correctly handle bfloat16 input and output from jax2tf functions. (#3652)

TF and JAX have different NumPy dtypes for bfloat16 so we need to be careful to
use the right version. I think there are a few other cases in jax2tf where we
should be using `to_tf_dtype` rather than passing `v.dtype` directly into tf
ops (e.g. I am a bit surprised to only update convert_element_type_p), however I
think a follow up adding tests for those cases would be best.

gnecula committed:
Refactored host_callback to use the C++ runtime. (#3644)

* Refactored host_callback to use the C++ runtime.

* The new runtime makes it unnecessary to start the outfeed_receiver
  in the user's code
* We don't need msgpack anymore
* There is an interaction between host_callback and using lax.outfeed.
  I am trying to solve this by (a) making host_callback_test stop the
  outfeed receiver on finish and infeed_test on start, and (b)
  telling pytest-xdist to run all the tests from one file into
  a single worker.

joschkabraun committed:
Merge branch 'numpy-setxor1d' of https://github.com/joschkabraun/jax into numpy-setxor1d

jakevdp commented:
Hi – looks like something is strange with the rebase. If you need help getting your branch back in order, please let me know what commands you ran.

jakevdp commented:
Since there has been no action here for a while, I'm going to close this. Feel free to re-open if you'd like to re-visit this contribution! Thanks
